WalletWasabi/Crypto/Groups/Generators.cs

```diff
@@ -63,32 +58,15 @@ public static class Generators
 		/// </summary>
 		public static GroupElement Gs { get; } = FromText("Gs");
 
-		public static GroupElement[] GetNegateGh2i(int rangeProofWidth)
-		{
-			lock (RangeProofWidthNegateGh2iPairsLock)
-			{
-				if (!RangeProofWidthNegateGh2iPairs.TryGetValue(rangeProofWidth, out var negateGh2i))
-				{
-					var negatedGh = Gh.Negate();
-					negateGh2i = GetPowersOfTwo(rangeProofWidth).Select(b => b * negatedGh).ToArray();
-					RangeProofWidthNegateGh2iPairs.Add(rangeProofWidth, negateGh2i);
-				}
-				return negateGh2i;
-			}
-		}
+		/// <summary>
+		/// Scalars corresponding to 2^i, used in range proofs.
+		/// </summary>
+		public static Scalar[] ScalarPowersOfTwo { get; } = Enumerable.Range(0, 255).Select(i => Scalar.Zero.CAddBit((uint)i, 1)).ToArray();
 
-		public static Scalar[] GetPowersOfTwo(int rangeProofWidth)
-		{
-			lock (RangeProofWidthPowersOfTwoPairsLock)
-			{
-				if (!RangeProofWidthPowersOfTwoPairs.TryGetValue(rangeProofWidth, out var powerOfTwo))
-				{
-					powerOfTwo = Enumerable.Range(0, rangeProofWidth).Select(i => Scalar.Zero.CAddBit((uint)i, 1)).ToArray();
-					RangeProofWidthPowersOfTwoPairs.Add(rangeProofWidth, powerOfTwo);
-				}
-				return powerOfTwo;
-			}
-		}
+		/// <summary>
+		/// Generators corresponding to -(2^i) * Gh, used in range proofs.
+		/// </summary>
+		public static GroupElement[] NegatedGhPowersOfTwo { get; } = ScalarPowersOfTwo.Select(b => b.Negate() * Gh).ToArray();
```

Comment:
Note that there's still implicit mutable state here because internally `Scalar * GE` returns a `GEJ`, and our `GroupElement` will compute the affine representation from the Jacobian one using a `Lazy` since that is expensive (requires modular inverses).

I was considering doing this eagerly for these generators, but `Lazy` is thread safe by default and the inversion will only happen once per generator, and `GroupElement` is a reference type this should be shared, so in the end I figured this way is cleaner/simpler,